include/cuda/memory_resource

```diff
@@ -0,0 +1,549 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the CUDA Toolkit, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_MEMORY_RESOURCE
+#define _CUDA_MEMORY_RESOURCE
+
+// clang-format off
+/*
+    memory_resource synopsis
+namespace cuda {
+namespace mr {
+template <class Resource>
+concept resource = equality_comparable<Resource>
```

Comment:
Why equality_comparable?

Reply:
that was one of the requirements given

Reply:
I don't think this requirement makes much sense. We have no way of comparing two resource refs after they've been type-erased, because there's no common base class - and without that, we're limited to simple `this` pointer equality. We can compare common vtable entries and if they match, call `reinterpret_cast` - but this may be more conservative than `operator==`, leading to unexpected behavior that's quite hard to understand for the user not familiar with the internal working of `resource_ref`. To that end, it's perhaps cleaner to implement the equality of `resource_ref` as a simple `return __object == other.__object`.

Reply:
> We have no way of comparing two resource refs after they've been type-erased, because there's no common base class - and without that, we're limited to simple this pointer equality.

That's not true. We can compare the vtable pointers for equality and this will tell us if two resource_refs refer to the same concrete type. If so, then we can invoke the stored `operator==` that casts the `void*` to the concrete type.

Reply:
Comparing the void pointers also doesn't work because of EBO / no_unique_address. For std::memory_resource this is never a concern, because of inheritance with virtual functions that necessarily makes all resource types at least one pointer, but we can very easily have empty resources so it's absolutely possible to have two different resources share the same address, so comparing the type erased pointer doesn't work.

What the PR does is the correct way to handle comparison operations for this formulation of the feature.